src/utils/detached-dom-stream.js

```diff
@@ -0,0 +1,63 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {userAssert} from '../log';
+
+export class DetachedDomStream {
+  /**
+   * @param {!Window} win
+   * @param {function(!Document):void} onChunk
+   * @param {function(!Document):void} onEnd
+   */
+  constructor(win, onChunk, onEnd) {
+    /** @const @private {function(!Document):void} */
+    this.onChunk_ = onChunk;
+
+    /** @const @private {function(!Document):void} */
+    this.onEnd_ = onEnd;
+
+    /** @const @private {!Document} */
+    this.detachedDoc_ = win.document.implementation.createHTMLDocument('');
+    this.detachedDoc_.open();
+
+    /** @private {boolean} */
+    this.eof_ = false;
+  }
+
+  /**
+   * Write chunk into detached doc, and call given chunk cb.
+   * @public
+   * @param {string} chunk
+   */
+  write(chunk) {
+    userAssert(!this.eof_, 'Detached doc already closed.');
```

Comment:
IMHO this should be a `devAssert`. Also, what happens when you write into a document you called a `close()` on? If it fails, we simply remove `this.eof_` and this check.

Reply:
calling `write()` again reopens, and destroys the old doc. Changed to `devAssert`.